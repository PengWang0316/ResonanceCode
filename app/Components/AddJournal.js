import React, { Component } from "react";
import Util from "../apis/Util";
import Loading from "./Loading"
import JournalContent from "./JournalContent";
import ReadingSearchAndList from "./ReadingSearchAndList";
import { isLogin } from "../apis/LoginApi";
import DatabaseApi from "../apis/DatabaseApi";
import PropTypes from "prop-types";

class AddJournal extends Component {

  componentWillMount(){
    // Initial variables
    this.readings = this.props.journalData ? this.props.journalData.readingIds : {}; // keep which readings should be attached on. Format is like {readingId: null}
    this.contents = {}; // Keep content keys and content
    this.contentKeyIndex = 0; // Generate keys for different contents
    this.contentIndexs = {}; // Use to track the index of content component in the array. Delete function needs it. The format is {contentKey: index}
    this.jounalId = this.props.journalData ? this.props.journalData._id : null; // Keeping journal id for update
    this.userId = isLogin(document).userid;

    this.state={
      isWriting: false,
      journalDate: this.props.journalData ? Util.getDateString(this.props.journalData.date) : Util.getCurrentDateString(),
      isDateCorrect: true,
      isEmptyReading: this.props.journalData ? false : true,
      contentComponentArray: [], // keep content component
      addJournalContent: "overview",
      pingPongState: this.props.journalData ? this.props.journalData.ping_pong_state : "Neutral"
    };

    console.log("Journal before alter: ", this.props.journalData);
    // If journal data exsit, get content
    if(this.props.journalData){
      // this.oldContentKeys = []; // keep original content keys
      this.oldReadingIds = Object.keys(this.readings); // keep original reading ids

      // delete unnecessary properties from journal object
      let journal = this.props.journalData;
      this.journalUserId = journal.user_id;
      delete journal.date;
      delete journal.user_id;
      delete journal.ping_pong_state;
      delete journal._id;
      delete journal.readingIds;
      let keyRegExp = /-(\d+)$/; // The regular expression for subtract suffixs
      console.log("Journal after alter: ", this.props.journalData);
      Object.keys(journal).map((key)=>{
        // this.oldContentKeys.push(key); // keeping the exsit keys in an array for delete function
        // handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared)
        // Also have to save the largest key number and use it to countine
        // journalData property is like {'overview-0': 'test overview', 'overview-0-isShared': true}
        // if key is isShared do not do anything
        let matchResult = key.match(keyRegExp);
        if (matchResult){
          if (this.contentKeyIndex < matchResult[1]) this.contentKeyIndex = matchResult[1]; // keep the max index number and user can contiune add new content
          this.handleAddContentClick(journal[key], key.replace(keyRegExp, ""), key, matchResult[1], journal[`${key}-isShared`]);
          // console.log("isShared: ",journal[`${key}-isShared`]);
        }
      });
      this.contentKeyIndex++;
    }

  }

  handleChange(event, element){
    let newState={};
    newState[element]=event.target.value;
    if(element=="journalDate") newState.isDateCorrect = Util.matchDateFormat(newState[element]);
    this.setState(newState);
  }

  handleChangeCallback(contentKey, contentText){
    console.log(contentKey, contentText);
    // update content in state
    this.contents[contentKey]=contentText;
  }

  handleDeleteContentCallback(contentKey){
    // Remove content from state
    if(this.contents.hasOwnProperty(contentKey)){
      delete this.contents[contentKey];
      delete this.contents[`${contentKey}-isShared`];
    }
    // Remove content component
    delete this.contentIndexs[contentKey];
    this.setComponentToStateArray();
    // this.state.contentComponentArray.splice(this.contentIndexs[contentKey],1);
    // this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSharedBoxChangeCallback(contentKey, isShared){
    this.contents[`${contentKey}-isShared`] = isShared;
  }

  handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared){
    // put component in array in order to show
    newContentKey = newContentKey ? newContentKey : `${this.state.addJournalContent}-${this.contentKeyIndex}`;
    // put the component in the track object
    this.contentIndexs[newContentKey]=<JournalContent key={contentKeyIndex ? contentKeyIndex : this.contentKeyIndex++} newContent={addJournalContent ? addJournalContent : ""} newContentName={newContentName ? newContentName : this.state.addJournalContent} newContentKey={newContentKey} handleChangeCallback={(contentKey, contentText)=>{this.handleChangeCallback(contentKey, contentText);}} handleDeleteContentCallback={(contentKey)=>{this.handleDeleteContentCallback(contentKey);}} handleSharedBoxChangeCallback={(contentKey, isShared)=>{this.handleSharedBoxChangeCallback(contentKey, isShared)}} isShared={isShared ? true : false} />;
    this.setComponentToStateArray();
    // this.setState({contentComponentArray: this.state.contentComponentArray});

    // also have to put content title to state in order to track content and update
    this.contents[newContentKey] = addJournalContent ? addJournalContent : "";
    this.contents[`${newContentKey}-isShared`] = isShared ? true : false;
  }

  setComponentToStateArray(){
    // transfor object to array in order to display
    this.state.contentComponentArray=[];
    Object.keys(this.contentIndexs).map((element)=>{
      this.state.contentComponentArray.push(this.contentIndexs[element]);
    });
    this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSubmit(event, isUpdate){
    event.preventDefault();
    // console.log("submit: contents:",this.contents);
    console.log("journal date:",this.state.journalDate);
    // Assemble a journal object for save
    let readingIdArray = [];
    Object.keys(this.readings).map((key)=>{readingIdArray.push(key)});
    let journal = Object.assign({_id:this.jounalId, date: new Date(this.state.journalDate), ping_pong_state: this.state.pingPongState, user_id: this.userId, readingIds: readingIdArray}, this.contents);

    if(isUpdate){
      console.log("submit: journal:", journal);
      // console.log("oldContentKeys:", this.oldContentKeys);
      console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      let deleteContents = [];
      let deleteReadingIds = [];
      // this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      this.oldReadingIds.map((element)=>{if(!this.readings.hasOwnProperty(element)) deleteReadingIds.push(element);});
      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      console.log("delete contents: ",deleteContents);
      console.log("delete reading ids : ",deleteReadingIds);
      DatabaseApi.updateJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }else{
      DatabaseApi.createJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }
    /**/
    this.setState({isWriting: true});
    console.log("submit: journal:", journal);
  }

  /* For ReadingSearchAndList callback */
  handleAttachReadingCallback(readingId){
    this.readings[readingId] = null;
    this.setState({isEmptyReading: false});
  }

  handleDetachAttachReadingCallback(readingId){
    delete this.readings[readingId];
    this.setState({isEmptyReading: Object.keys(this.readings).length===0});
  }

  handleCancel(){
    this.props.history.push("/reading");
  }

  handleDelete(){
    console.log("Delete journal!");
  }

  render(){
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        <div className="titleDiv">{this.props.journalData ? "Update journal for readings" : "Add a new journal for readings"}</div>
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event, this.props.journalData);}}>

          <div className="text-right bottom-btn-div">
            {(!this.props.journalData || this.userId == this.journalUserId) &&
              <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
            }
            {(this.props.journalData && this.userId == this.journalUserId) &&
              <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
            }
            <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
          </div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">

                <input className={this.state.isDateCorrect?"form-control":"form-control form-control-warning"} type="text" placeholder="mm/dd/yyyy" id="journalDate" value={this.state.journalDate} onChange={(event)=>{this.handleChange(event, "journalDate")}} />
                {!this.state.isDateCorrect && <span className="glyphicon glyphicon-warning-sign form-control-feedback form-control-warning-span"></span>}

              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            {/* New content goes here */}
            {this.state.contentComponentArray}

            {/* Add content button and drop list */}
            <div className="row addJournalContentDiv">

              <div onClick={()=>{this.handleAddContentClick();}} className="addJournalContentBtnDiv col-sm-6"><i className="fa fa-plus-square" /> Add one content for your journal</div>
              <div className="col-sm-6">
                <select  className="form-control" onChange={(event)=>{this.handleChange(event,"addJournalContent");}}>
                  <option value="overview">Overview</option>
                  <option value="internal_environment">Internal Environment</option>
                  <option value="relational_environment">Relational Environment</option>
                  <option value="physical_environment">Physical Environment</option>
                  <option value="creative_vector">Creative Vector</option>
                  <option value="synchronicities">Synchronicities</option>
                  <option value="dreams">Dreams</option>
                  <option value="other">Other</option>
                </select>

              </div>

            </div>


            {/*  Start reading search function */}
            <ReadingSearchAndList readings={this.props.journalData ? this.readings : null} attachReadingCallback={(readingId)=>{this.handleAttachReadingCallback(readingId);}} detachReadingCallback={(readingId)=>{this.handleDetachAttachReadingCallback(readingId);}} />

              <div className="text-right bottom-btn-div">
                {(!this.props.journalData || this.userId == this.journalUserId) &&
                  <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
                }
                {(this.props.journalData && this.userId == this.journalUserId) &&
                  <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
                }
                <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
              </div>
          </form>

        </div>
    );
  }


}
AddJournal.propTypes={
  journalData: PropTypes.object,
  history: PropTypes.object
};
export default AddJournal;
