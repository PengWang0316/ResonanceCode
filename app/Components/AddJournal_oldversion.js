import React, { Component } from "react";
import QueryString from "query-string";
import PropTypes from "prop-types";
import Util from "../apis/Util";
// import api from "../apis/api";
import DatabaseApi from "../apis/DatabaseApi";
import Loading from "./Loading";
import LoginApi from "../apis/LoginApi";

class AddJournal_oldversion extends Component{

  componentWillMount(){

    if(this.props.location){
      let queryInfo = QueryString.parse(this.props.location.search);
      this.readingId = queryInfo.readingId;
      this.readingName = queryInfo.readingName;
      this.readingDate = Util.getDateString(queryInfo.readingDate);
      // console.log("AddJournal paper readingId, readingName, and readingDate:", `Id: ${this.readingId} Name: ${this.readingName} Date: ${this.readingDate}`);
    }
    this.userId=LoginApi.isLogin(document).userid;

    let journalData = this.props.journalData;
    if(journalData){
      this.journalId=journalData.journalId;
      // this.readingDate=journalData.readingDate;
      this.readingName=journalData.readingName;
    }
    // console.log("AddJournal page:",journalData);
    // setting state
    this.state={
      isWriting: false,
      journalDate: journalData?Util.getDateString(journalData.date):Util.getCurrentDateString(),
      pingPongState: journalData?journalData.pingPongState:"Neutral",
      overview: journalData?journalData.overview:"",
      internalEnvironment: journalData?journalData.internalEnvironment:"",
      relationalEnvironment: journalData?journalData.relationalEnvironment:"",
      physicalEnvironment: journalData?journalData.physicalEnvironment:"",
      creativeVector: journalData?journalData.creativeVector:"",
      synchronicities: journalData?journalData.synchronicities:"",
      dreams: journalData?journalData.dreams:"",
      journalId: journalData?journalData.journalId:"",
      userId: journalData?journalData.userId:"",
      readingId: journalData?journalData.readingId:""
    };
    // if(this.journalId) this.setState(this.state);
  }

  componentDidMount(){
    // console.log("did mount");
    if(this.journalId) this.setState(this.state);
  }

  handleChange(event, element){
    // console.log(event.target.value);
    let newState={};
    newState[element]=event.target.value;
    this.setState(newState);
    // console.log(newState);
  }

  handleSubmit(event){
    event.preventDefault();
    // console.log("**********submit************");

      this.setState({isWriting: true});
      // console.log("submit:",this.state);
      // assemble object for creating
      let journalObject = {
        user_id: this.userid,
        reading_id: this.readingId?this.readingId:this.state.readingId,
        date: new Date(this.state.journalDate),
        overview: this.state.overview,
        internal_environment: this.state.internalEnvironment,
        relational_environment: this.state.relationalEnvironment,
        physical_environment: this.state.physicalEnvironment,
        creative_vector: this.state.creativeVector,
        synchronicities: this.state.synchronicities,
        dreams: this.state.dreams,
        ping_pong_state:this.state.pingPongState
      };
      // console.log("journalId", this.journalId);
      if(this.journalId) DatabaseApi.updateJournal(journalObject, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
      else DatabaseApi.createJournal(journalObject).then((reault)=>{this.props.history.push("/reading")});

  }

  handleDelete(event){
    event.preventDefault();
    // console.log("**********delete************");
    DatabaseApi.deleteJournal(this.state.readingId, this.state.journalId).then((reault)=>{this.props.history.push("/reading")});
  }

  render(){
    let titleDiv;
    if (this.readingId) titleDiv = <div><span className="rcTitle">Add a new journal to your reading: </span>{this.readingName}  ({this.readingDate})</div>;
      else titleDiv = <div><span className="rcTitle">Update {this.readingName}</span></div>;
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        {titleDiv}
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event);}}>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button type="button" className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">
                <input className="form-control" type="text" value={this.state.journalDate} id="journalDate" onChange={(event)=>{this.handleChange(event,"journalDate");}} />
              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="overview" className="col-form-label">Overview</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.overview} placeholder="Overview..." id="overview" onChange={(event)=>{this.handleChange(event,"overview");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="internalEnvironment" className="col-form-label">Internal Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.internalEnvironment} placeholder="Internal Environment..." id="internalEnvironment" onChange={(event)=>{this.handleChange(event,"internalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="relationalEnvironment" className="col-form-label">Relational Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.relationalEnvironment} placeholder="Relational Environment..." id="relationalEnvironment" onChange={(event)=>{this.handleChange(event,"relationalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="physicalEnvironment" className="col-form-label">Physical Environment</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.physicalEnvironment} placeholder="Physical Environment..." id="physicalEnvironment" onChange={(event)=>{this.handleChange(event,"physicalEnvironment");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="creativeVector" className="col-form-label">Creative Vector</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.creativeVector} placeholder="Creative Vector..." id="creativeVector" onChange={(event)=>{this.handleChange(event,"creativeVector");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="synchronicities" className="col-form-label">Synchronicities</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.synchronicities} placeholder="Synchronicities..." id="synchronicities" onChange={(event)=>{this.handleChange(event,"synchronicities");}} />
              </div>
            </div>

            <div className="form-group form-div">
              <label htmlFor="dreams" className="col-form-label">Dreams</label>
              <div>
                <textarea className="form-control" rows="3" type="text" value={this.state.dreams} placeholder="Dreams..." id="dreams" onChange={(event)=>{this.handleChange(event,"dreams");}} />
              </div>
            </div>

          <div className="text-right bottom-btn-div">{(!this.readingId && this.state.userId==this.userId) && <button className="btn btn-danger btn-del" onClick={(event)=>{this.handleDelete(event);}}>Delete</button>}{(this.readingId || this.state.userId==this.userId) && <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.overview.length>0)}>{this.readingId?"Submit":"Update"}</button>}</div>

        </form>

      </div>
    );
  }

}
AddJournal_oldversion.propTypes={
  journalData: PropTypes.object
};
export default AddJournal_oldversion;
