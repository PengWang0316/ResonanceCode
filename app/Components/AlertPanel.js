import React from 'react';

import styles from '../styles/AlertPanel.module.css';

const AlertPanel = ({
  id = '', type = 'success', children = '', style = {}
}) => (
  <div id={id} className={`alert alert-${type} mt-3 ${styles.alertPanel}`} style={style} role="alert">
    {children}
  </div>
);
export default AlertPanel;
