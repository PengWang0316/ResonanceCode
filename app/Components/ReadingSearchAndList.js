import React, { Component } from "react";
import PropTypes from "prop-types";
import { getReadingBasedOnName } from "../apis/DatabaseApi";
import { isLogin } from "../apis/LoginApi";

class ReadingSearchAndList extends Component{

  componentWillMount(){
    console.log("ReadingSearchAndList: ", this.props);
    this.readingIndexTracker = {}; // Tracking reading index in the array (delete function needs)
    this.readingIndex = 0;
    this.searchFunction = null; // Keeping search function
    this.userId = isLogin(document).userid;
    this.state = {
      searchReading: "", // tracking search input text
      readingArray: [], // components for readings
      searchResults: [] // Keeping search results
    };

    /*putting readings' pingPongStates to state if it is updating
      The format is like {readingId: pingPongState}
    */
    this.pingPongStates = this.props.pingPongStates;

    // console.log("this.props.readings", this.props.readings);
    // if readings has already exsited, put them in the list in order to update
    if(this.props.readings){
      Object.keys(this.props.readings).map((readingId)=>{
        this.handleAddReading(this.props.readings[readingId], readingId, this.readingIndex++);
      });
    }

  }

  handleChange(event, element){
    let newState={};
    let keyWord = event.target.value;
    newState[element] = keyWord;
    // start to search
    if (keyWord.length > 2) this.searchKeyWord(keyWord);
    this.setState(newState);
  }

  /*
   *  Considering the performance, setting up a timeout to delay search api call for 1s.
   *  If somebody type fast, it can save serval database operation
  */
  searchKeyWord(keyWord){
    // console.log("start to search: ", keyWord);
    if(this.searchFunction) clearTimeout(this.searchFunction);
    this.searchFunction = setTimeout(()=>{
      // console.log(`${keyWord} api call`);
      getReadingBasedOnName(keyWord, this.userId).then((result)=>{
        this.state.searchResults=[];
        let keyWordExpression = new RegExp(`(.*)(${keyWord})(.*)`,"i");
        // console.log("result:", result);
        result.data.map((element, index)=>{
          // Format the new
          let nameResult = element.reading_name.match(keyWordExpression);
          // console.log(nameResult);
          this.state.searchResults.push(<div onClick={()=>{this.handleResultClick(element._id, element.reading_name);}} className="readingSearchItem" key={index}>{nameResult[1]}<span className="matchKeyword">{nameResult[2]}</span>{nameResult[3]}</div>);
        });

        if (this.state.searchResults.length===0) this.state.searchResults.push(<div key="noResult" className="readingSearchItem">No Result</div>);
        this.setState({searchResults: this.state.searchResults});
      });

      /*this.state.searchResults.push(<div key={keyWord}>{keyWord}</div>);
      this.setState({searchResults: this.state.searchResults});*/
    }, 1000);
  }

  handleResultClick(readingId, readingName){
    // console.log(readingId, readingName);
    this.setState({searchResults: [], searchReading: ""});
    this.handleAddReading(readingName, readingId, this.readingIndex++);
  }

  handlePingPongStateChange(readingId, event){
    // console.log(readingId, event.target.value);
    // this.setState([readingId]: event.target.value);
    this.props.handlePingpongstateChangeCallback(readingId, event.target.value);
  }

  handleAddReading(readingName, readingId, readingIndex){
    // console.log("key:", readingIndex)
    // Putting reading component in the object
    // let pingPongState = this.state[readingId];
    this.readingIndexTracker[readingIndex] =
    <div key={readingIndex}>

      <div className="row readingListNameDiv">
        <div className="col-xs-10">{readingName}</div>
        <div className="col-xs-2 readingListDeletIcon" onClick={()=>{this.handleDelete(readingId, readingIndex);}}><i className="fa fa-trash delete-icon" /></div>
      </div>

      {/*<label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>*/}
      <div className="row pingPongStateDiv">
        <div className="col-xs-1">&#8627;</div>
        <div className="col-xs-10">
          <select id="pingPongState" className="form-control" defaultValue={this.pingPongStates[readingId]?this.pingPongStates[readingId]:"Neutral"} onChange={(event)=>{this.handlePingPongStateChange(readingId, event);}}>
            <option value="Neutral">Neutral</option>
            <option value="Reading/Call">Reading/Call</option>
            <option value="Impact absorbing">Impact absorbing</option>
            <option value="Impact complete">Impact complete</option>
            <option value="Response">Response</option>
            <option value="Outward moving">Outward moving</option>
            <option value="Outward complete">Outward complete</option>
          </select>
        </div>
      </div>

    </div>;


    this.setReadingToStateArray();
    this.props.attachReadingCallback(readingId);

  }

  handleDelete(readingId, readingIndex){
    // console.log("delete reading:", readingId, readingIndex);
    // this.state.readingArray.splice(this.readingIndexTracker[readingIndex], 1);
    delete this.readingIndexTracker[readingIndex];
    this.setReadingToStateArray();
    this.props.detachReadingCallback(readingId);
    // console.log("array:",this.state.readingArray);
  }

  setReadingToStateArray(){
    // transfor object to array in order to display
    this.state.readingArray=[];
    Object.keys(this.readingIndexTracker).map((element)=>{
      this.state.readingArray.push(this.readingIndexTracker[element]);
    });
    this.setState({readingArray: this.state.readingArray});
  }

  render(){
    return(
      <div className="row addJournalContentDiv">

        {/* Search input */}
        <div className="col-sm-5">
          <div><label htmlFor="searchReading" className="col-form-label">Search and pick up readings</label></div>
          <div>
            <input className="form-control" type="text" placeholder="Type reading name here..." id="searchReading" value={this.state.searchReading} onChange={(event)=>{this.handleChange(event, "searchReading")}} />
          </div>

          {/* showing searching result */}
          {this.state.searchResults.length!==0 &&
            <div className="readingSearchResultDiv">
              {this.state.searchResults}
            </div>
          }
        </div>

        {/* Reading list */}
        <div className="col-sm-7">
          <div className="font-bold">Reading List (This journal will attach on)</div>
          <div className="container-fluid readingListDiv">
            {this.state.readingArray}
          </div>
        </div>

      </div>
    );
  }

}
ReadingSearchAndList.propTypes = {
  detachReadingCallback: PropTypes.func.isRequired,
  attachReadingCallback: PropTypes.func.isRequired,
  readings: PropTypes.object
};
export default ReadingSearchAndList;
