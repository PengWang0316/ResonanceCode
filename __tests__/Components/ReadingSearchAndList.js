import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { ReadingSearchAndList } from '../../app/Components/ReadingSearchAndList';

jest.mock('../../app/Components/ReadingSearchResult', () => 'ReadingSearchResult');
jest.mock('../../app/Components/ReadingSearchItem', () => 'ReadingSearchItem');
jest.mock('../../app/Components/AllReadingList', () => 'AllReadingList');

describe('ReadingSearchAndList test', () => {
  const defaultProps = {
    existReadings: { readingIdA: {} },
    pingPongStates: { readingIdA: 'pingPongStateA' },
    searchReadings: [{ _id: 'readingA', reading_name: 'readingNameA' }],
    allReadingList: [],
    // readingsAmount: PropTypes.number,
    attachReadingCallback: jest.fn(),
    detachReadingCallback: jest.fn(),
    handlePingpongstateChangeCallback: jest.fn(),
    fetchReadingBasedOnName: jest.fn(),
    fetchAllReadingList: jest.fn(),
    fetchReadingsAmount: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<ReadingSearchAndList {...props} />);

  test('ReadingSearchAndList snapshot', () => expect(renderer.create(<ReadingSearchAndList {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('initial state and functions were called', () => {
    const mockFunc = jest.fn();
    const component = getShallowComponent({ ...defaultProps, fetchReadingsAmount: mockFunc });
    expect(component.state('searchReading')).toBe('');
    expect(component.state('readingArray').length).toBe(1);
    expect(component.state('searchResults')).toEqual([]);
    const {
      readingIndexTracker, readingIndex, searchFunction, pingPongStates
    } = component.instance();
    expect(Object.keys(readingIndexTracker).length).toBe(1);
    expect(readingIndex).toBe(1);
    expect(searchFunction).toBeNull();
    expect(pingPongStates).toBe(component.instance().props.pingPongStates);
    expect(defaultProps.handlePingpongstateChangeCallback).toHaveBeenLastCalledWith('readingIdA', 'pingPongStateA');
    expect(defaultProps.attachReadingCallback).toHaveBeenLastCalledWith('readingIdA');
    expect(mockFunc).toHaveBeenCalledTimes(1);
  });

  test('componentWillReceiveProps', () => {
    const component = getShallowComponent();
    expect(component.find('ReadingSearchResult').length).toBe(0);
    component.setProps({
      searchReadings: [
        ...defaultProps.searchReadings,
        { _id: 'readingB', reading_name: 'readingNameB' }
      ]
    });
    expect(component.find('ReadingSearchResult').length).toBe(2);
  });

  test('componentWillReceiveProps with zero search result', () => {
    const component = getShallowComponent();
    component.setProps({ searchReadings: [] });
    expect(component.find('ReadingSearchResult').length).toBe(0);
    expect(component.state('searchResults')).toEqual([<div key="noResult" className="readingSearchItem">No Result</div>]);
  });

  test('handleChange and searchKeyWord', () => {
    jest.useFakeTimers();
    const component = getShallowComponent();
    component.find('input').simulate('change', { target: { value: '12', id: 'newId' } });
    expect(component.state('searchResults')).toEqual([]);
    expect(component.state('newId')).toBe('12');
    expect(clearTimeout).not.toHaveBeenCalled();
    expect(setTimeout).not.toHaveBeenCalled();
    expect(defaultProps.fetchReadingBasedOnName).not.toHaveBeenCalled();
    component.instance().searchFunction = jest.fn();
    component.find('input').simulate('change', { target: { value: '123', id: 'newId' } });
    expect(clearTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenCalledTimes(1);
    jest.runAllTimers();
    expect(defaultProps.fetchReadingBasedOnName).toHaveBeenLastCalledWith('123');
    expect(component.instance().keyWord).toBe('123');
  });

  test('handleResultClickCallback', () => {
    const component = getShallowComponent();
    component.instance().readingIndex = 1;
    component.find('AllReadingList').prop('handleClick')('readingIdAAA', 'readingNameAAA');
    expect(component.state('searchResults')).toEqual([]);
    expect(component.state('searchReading')).toBe('');
    expect(component.instance().readingIndex).toBe(2);
  });

  test('handlePingPongStateChangeCallback', () => {
    const component = getShallowComponent();
    component.find('ReadingSearchItem').prop('handlePingPongStateChangeCallback')('readingIdAA', 'optionValue');
    expect(defaultProps.handlePingpongstateChangeCallback).toHaveBeenLastCalledWith('readingIdAA', 'optionValue');
  });

  test('handleDeleteCallback', () => {
    const component = getShallowComponent();
    const mockFunc = jest.fn();
    component.instance().setReadingToStateArray = mockFunc;
    component.instance().readingIndexTracker.readingIndex = {};
    component.find('ReadingSearchItem').prop('handleDeleteCallback')('readingIdAA', 'readingIndex');
    expect(component.instance().readingIndexTracker.readingIndex).toBeUndefined();
    expect(mockFunc).toHaveBeenCalledTimes(1);
    expect(defaultProps.detachReadingCallback).toHaveBeenLastCalledWith('readingIdAA');
  });

  test('handleShowReadingListClick', () => {
    const component = getShallowComponent();
    const mockFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ modal: mockFunc });
    component.setProps({ allReadingList: [] });
    component.find('button').simulate('click');
    expect(defaultProps.fetchAllReadingList).toHaveBeenLastCalledWith(0);
    expect(window.$).toHaveBeenLastCalledWith('#readingListModal');
    expect(mockFunc).toHaveBeenLastCalledWith('toggle');
  });

  test('preventSubmit', () => {
    const component = getShallowComponent();
    const mockFunc = jest.fn();
    component.find('input').simulate('keyPress', { charCode: 13, preventDefault: mockFunc });
    expect(mockFunc).toHaveBeenCalledTimes(1);
  });

  // test('handleChange', () => {
  //   const component = getShallowComponent();
  //   component.find('input').simulate('change', { target: { id: 'newId', value: 'newValue' } });
  //   expect(component.state('searchResults')).toEqual([]);
  //   epect()
  // });
});
