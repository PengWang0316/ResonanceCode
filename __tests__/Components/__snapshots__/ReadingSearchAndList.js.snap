// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingSearchAndList test ReadingSearchAndList snapshot 1`] = `
<div
  className="row addJournalContentDiv"
>
  <div
    className="col-sm-5"
  >
    <div>
      <label
        className="col-form-label font-weight-bold"
        htmlFor="searchReading"
      >
        Search and select readings
      </label>
    </div>
    <div>
      <input
        className="form-control"
        id="searchReading"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Type reading name here..."
        type="text"
        value=""
      />
      <small
        className="form-text text-muted mb-2"
        id="searchReadingHelp"
      >
        Skip if this is a pre-reading entry.
      </small>
      <div>
        <button
          className="btn btn-primary btn-sm"
          onClick={[Function]}
          type="button"
        >
          Show all reading list
        </button>
      </div>
    </div>
  </div>
  <div
    className="col-sm-7"
  >
    <div
      className="font-weight-bold"
    >
      Reading List (This journal can be attached to multiple readings.)
    </div>
    <div
      className="container-fluid readingListDiv"
    >
      <ReadingSearchItem
        handleDeleteCallback={[Function]}
        handlePingPongStateChangeCallback={[Function]}
        pingPongStates={
          Object {
            "readingIdA": "pingPongStateA",
          }
        }
        readingId="readingIdA"
        readingIndex={0}
        readingName={Object {}}
      />
    </div>
  </div>
  <AllReadingList
    handleClick={[Function]}
  />
</div>
`;
